### PR TITLE
[FLINK-5907] [java api] Fix trailing empty fields in CsvInputFormat

### PR DESCRIPTION
**Pull request by KurtYoung**
If there are 3 fields with field delimiter ",", both these two line should be parsed successfully:
aaa,bbb,,
aaa,bbb,

### CODE REVIEW

**Comment by fpompermaier** [comment]
I've tested your PR against my original CSV that was causing the problem and it works now! Thanks for this fix

File: flink-java/src/main/java/org/apache/flink/api/java/io/RowCsvInputFormat.java [hunk 1]
```diff
@@ -197,6 +197,14 @@ protected boolean parseRecord(Object[] holders, byte[] bytes, int offset, int nu
 			if (startPos < 0) {
 				throw new ParseException(String.format("Unexpected parser position for column %1$s of row '%2$s'",
 					field, new String(bytes, offset, numBytes)));
+			} else if (startPos == limit
+					&& field != fieldIncluded.length - 1
+					&& !FieldParser.endsWithDelimiter(bytes, startPos - 1, fieldDelimiter)) {
+				if (isLenient()) {
```
Comment (fhueske) [inline flink-java/src/main/java/org/apache/flink/api/java/io/RowCsvInputFormat.java]: Add a comment: "We are at the end of the record, but not all fields have been read and the end is not a field delimiter indicating an empty last field."

File: flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java [hunk 2]
```diff
@@ -358,24 +358,27 @@ protected boolean parseRecord(Object[] holders, byte[] bytes, int offset, int nu
...
 				FieldParser<Object> parser = (FieldParser<Object>) this.fieldParsers[output];
 				Object reuse = holders[output];
 				startPos = parser.resetErrorStateAndParse(bytes, startPos, limit, this.fieldDelim, reuse);
 				holders[output] = parser.getLastResult();
-				
+
 				// check parse result
-				if (startPos < 0) {
+				if (startPos < 0 ||
+						(startPos == limit
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java]: Move this condition into an `else if` branch and give a more detailed error message (row to short).
Also add a comment that we read the whole records but that there are fields missing.

File: flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java [hunk 3]
```diff
@@ -392,12 +395,17 @@ protected boolean parseRecord(Object[] holders, byte[] bytes, int offset, int nu
 			else {
 				// skip field
 				startPos = skipFields(bytes, startPos, limit, this.fieldDelim);
-				if (startPos < 0) {
+				if (startPos < 0 ||
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java]: Same here as above

File: flink-java/src/test/java/org/apache/flink/api/java/io/CsvInputFormatTest.java [hunk 4]
```diff
@@ -400,7 +400,7 @@ public void readMixedQuotedStringFields() {
 	@Test
 	public void readStringFieldsWithTrailingDelimiters() {
 		try {
-			final String fileContent = "abc|-def|-ghijk\nabc|-|-hhg\n|-|-|-\n";
+			final String fileContent = "abc|-def|-ghijk\nabc|-|-hhg\n|-|-|-\n|-|-\nabc|-def\n";
```
Comment (KurtYoung) [inline flink-java/src/test/java/org/apache/flink/api/java/io/CsvInputFormatTest.java]: sounds good

File: flink-java/src/test/java/org/apache/flink/api/java/io/CsvInputFormatTest.java [hunk 5]
```diff
@@ -957,6 +964,42 @@ public void testPojoTypeWithPrivateField() throws Exception {
 	}
 
 	@Test
+	public void testPojoTypeWithTrailingEmptyFields() throws Exception {
+		File tempFile = File.createTempFile("CsvReaderPojoType", "tmp");
```
Comment (KurtYoung) [inline flink-java/src/test/java/org/apache/flink/api/java/io/CsvInputFormatTest.java]: changed

File: flink-java/src/test/java/org/apache/flink/api/java/io/RowCsvInputFormatTest.java [hunk 6]
```diff
@@ -311,7 +317,7 @@ public void readMixedQuotedStringFields() throws Exception {
 
 	@Test
 	public void readStringFieldsWithTrailingDelimiters() throws Exception {
-		String fileContent = "abc|-def|-ghijk\nabc|-|-hhg\n|-|-|-\n";
+		String fileContent = "abc|-def|-ghijk\nabc|-|-hhg\n|-|-|-\n|-|-\nabc|-def\n";
```
Comment (KurtYoung) [inline flink-java/src/test/java/org/apache/flink/api/java/io/RowCsvInputFormatTest.java]: ok

File: flink-java/src/main/java/org/apache/flink/api/java/io/RowCsvInputFormat.java (hunk 1 above)
Comment (KurtYoung) [inline flink-java/src/main/java/org/apache/flink/api/java/io/RowCsvInputFormat.java]: added

File: flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java (hunk 3 above)
Comment (KurtYoung) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java]: done

File: flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java (hunk 2 above)
Comment (KurtYoung) [inline flink-core/src/main/java/org/apache/flink/api/common/io/GenericCsvInputFormat.java]: done

**Comment by fhueske** [comment]
Thanks for the fast update @KurtYoung.
+1 to merge

**Comment by fhueske** [comment]
Merging

**Pull request closed without merging**